packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/LightCurveViewer/LightCurveViewerContainer.spec.js

```diff
@@ -1,20 +1,173 @@
 import React from 'react'
+import nock from 'nock'
+import sinon from 'sinon'
+import * as d3 from 'd3'
+import { zip } from 'lodash'
+import asyncStates from '@zooniverse/async-states'
 
 import LightCurveViewerContainer from './LightCurveViewerContainer'
+import mockData from './mockData.json'
+import { Factory } from 'rosie'
 
 let wrapper
 
-describe('Component > LightCurveViewerContainer', function () {
-  beforeEach(function () {
-    wrapper = shallow(<LightCurveViewerContainer.wrappedComponent />)
-  })
+const subject = Factory.build('subject', { locations: [
+  { 'application/json': 'http://localhost:8080/mockData.json' }
+]})
+
+const nextSubject = Factory.build('subject', { locations: [
+  { 'text/plain': 'http://localhost:8080/nextSubject.json' }
+] })
+
+const nextSubjectJSON = { x: [1, 2], y: [3, 4] }
 
+const imageSubject = Factory.build('subject')
+
+const failSubject = Factory.build('subject', {
+  locations: [
+    { 'application/json': 'http://localhost:8080/failure.json' }
+  ]
+})
+
+describe('Component > LightCurveViewerContainer', function () {
   it('should render without crashing', function () {
+    wrapper = shallow(<LightCurveViewerContainer.wrappedComponent />)
     expect(wrapper).to.be.ok
   })
 
   it('should render null if there is no subject prop', function () {
-    expect(wrapper.type()).to.be.null
+    wrapper = shallow(<LightCurveViewerContainer.wrappedComponent />)
+    expect(wrapper.html()).to.be.null
+  })
+
+  it('should mount with an initialized state', function () {
+    wrapper = shallow(<LightCurveViewerContainer.wrappedComponent />, { disableLifecycleMethods: true })
+    const mockState = {
+      loading: asyncStates.initialized,
+      error: '',
+      dataExtent: { x: [], y: [] },
+      dataPoints: []
+    }
+    expect(wrapper.state()).to.eql(mockState)
+  })
+
+  describe('when there is a subject', function () {
+    const mockD3XExtent = d3.extent(mockData.x)
+    const mockD3YExtent = d3.extent(mockData.y)
+    const mockZipData = zip(mockData.x, mockData.y)
+    let cdmSpy
+    let nockScope
+
+    before(function () {
+      cdmSpy = sinon.spy(LightCurveViewerContainer.wrappedComponent.prototype, 'componentDidMount')
+      nockScope = nock('http://localhost:8080')
+        .persist(true)
+        .get('/mockData.json')
+        .reply(200, mockData)
+        .get('/failure.json')
+        .reply(404)
+        .get('/nextSubject.json')
+        .reply(200, nextSubjectJSON)
+
+    })
+
+    afterEach(function () {
+      cdmSpy.resetHistory()
+    })
+
+    after(function () {
+      cdmSpy.restore()
+      nock.cleanAll()
+      nockScope.persist(false)
+    })
+
+    it('should error if a json or text subject location file can\'t be found', function (done) {
+      wrapper = shallow(
+        <LightCurveViewerContainer.wrappedComponent subject={imageSubject} />
+      )
+
+      expect(cdmSpy).to.have.been.calledOnce
+      cdmSpy.returnValues[0].then(() => {
```

Comment:
I've opened #810 to try this out for `componentDidMount`.

Reply:
I did a bunch of reading about how to test an async `componentDidMount` and the maintainers seem to prefer having the promise returned to make it available for tests: https://github.com/airbnb/enzyme/issues/1587#issuecomment-400934989

Using `setTimeout` seems more brittle to me because it's arbitrary rather than explicitly using the resolve value from the Promise.

Reply:
If the exact timeout is the problem, I suppose you could set an explicit delay for the mock response with `nock.setDelay()` then expect state to change after that time.

My problem with using the promise is that you're changing the component code for the tests, so you're testing implementation, not behaviour (https://github.com/airbnb/enzyme/issues/1587#issuecomment-438876545). You've limited how much you can refactor the component, because you've tied the implementation of your tests to the implementation of `componentDidMount`. This might not be big deal, however.